// ngx-editor/esm2020/lib/editor.module.mjs
import { NgModule, InjectionToken } from '@angular/core';
import { CommonModule } from '@angular/common';
import { defaults as defaultLocals } from './Locals';
import { NgxEditorComponent } from './editor.component';
import { NgxEditorServiceConfig } from './editor-config.service';
import { NgxEditorService, provideMyServiceOptions } from './editor.service';
import { MenuModule } from './modules/menu/menu.module';
import { MenuComponent } from './modules/menu/menu.component';
import { ImageViewComponent } from './components/image-view/image-view.component';
import { FloatingMenuComponent } from './modules/menu/floating-menu/floating-menu.component';
import * as i0 from "@angular/core";
export const NGX_EDITOR_CONFIG_TOKEN = new InjectionToken('NgxEditorConfig');
const defaultConfig = {
    locals: defaultLocals,
};
export class NgxEditorModule {
    static forRoot(config = defaultConfig) {
        return {
            ngModule: NgxEditorModule,
            providers: [
                {
                    provide: NGX_EDITOR_CONFIG_TOKEN,
                    useValue: config,
                },
                {
                    provide: NgxEditorServiceConfig,
                    useFactory: provideMyServiceOptions,
                    deps: [NGX_EDITOR_CONFIG_TOKEN],
                },
            ],
        };
    }
    static forChild(config = defaultConfig) {
        return {
            ngModule: NgxEditorModule,
            providers: [
                {
                    provide: NGX_EDITOR_CONFIG_TOKEN,
                    useValue: config,
                },
                {
                    provide: NgxEditorServiceConfig,
                    useFactory: provideMyServiceOptions,
                    deps: [NGX_EDITOR_CONFIG_TOKEN],
                },
                NgxEditorService,
            ],
        };
    }
}
NgxEditorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: NgxEditorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxEditorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: NgxEditorModule, declarations: [NgxEditorComponent,
        ImageViewComponent], imports: [CommonModule,
        MenuModule], exports: [NgxEditorComponent,
        MenuComponent,
        FloatingMenuComponent] });
NgxEditorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: NgxEditorModule, providers: [], imports: [[
            CommonModule,
            MenuModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: NgxEditorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MenuModule,
                    ],
                    providers: [],
                    declarations: [
                        NgxEditorComponent,
                        ImageViewComponent,
                    ],
                    exports: [
                        NgxEditorComponent,
                        MenuComponent,
                        FloatingMenuComponent,
                    ],
                }]
        }] });
